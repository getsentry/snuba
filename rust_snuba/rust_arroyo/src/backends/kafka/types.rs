use rdkafka::message::OwnedHeaders;

#[derive(Clone)]
pub struct KafkaPayload {
    pub key: Option<Vec<u8>>,
    pub headers: Option<OwnedHeaders>,
    pub payload: Option<Vec<u8>>,
}
