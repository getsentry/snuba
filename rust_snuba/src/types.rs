use std::cmp::min;
use std::collections::BTreeMap;

use crate::strategies::clickhouse::batch::HttpBatch;

use chrono::{DateTime, Utc};
use rust_arroyo::backends::kafka::types::KafkaPayload;
use rust_arroyo::timer;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq)]
pub struct CommitLogEntry {
    pub offset: u64,
    pub orig_message_ts: DateTime<Utc>,
    pub received_p99: Vec<DateTime<Utc>>,
}

#[derive(Default, Debug, Clone)]
pub struct CommitLogOffsets(pub BTreeMap<u16, CommitLogEntry>);

impl CommitLogOffsets {
    fn merge(&mut self, other: CommitLogOffsets) {
        for (partition, other_entry) in other.0 {
            self.0
                .entry(partition)
                .and_modify(|entry| {
                    entry.offset = other_entry.offset;
                    entry.orig_message_ts = other_entry.orig_message_ts;
                    entry.received_p99.extend(&other_entry.received_p99);
                })
                .or_insert(other_entry);
        }
    }
}

#[derive(Clone, Debug, PartialEq, Default)]
pub struct CogsData {
    pub data: BTreeMap<String, u64>, // app_feature: bytes_len
}

impl CogsData {
    fn merge(&mut self, other: CogsData) {
        for (k, v) in other.data {
            self.data
                .entry(k)
                .and_modify(|curr| *curr += v)
                .or_insert(v);
        }
    }
}

#[derive(Debug, Clone)]
struct LatencyRecorder {
    sum_timestamps: f64,
    earliest_timestamp: u64,
    num_values: usize,
}

impl Default for LatencyRecorder {
    fn default() -> Self {
        LatencyRecorder {
            sum_timestamps: 0.0,
            earliest_timestamp: u64::MAX,
            num_values: 0,
        }
    }
}

impl From<DateTime<Utc>> for LatencyRecorder {
    fn from(value: DateTime<Utc>) -> Self {
        let value = value.timestamp_millis();
        LatencyRecorder {
            sum_timestamps: value as f64,
            earliest_timestamp: value as u64,
            num_values: 1,
        }
    }
}

impl LatencyRecorder {
    fn merge(&mut self, other: Self) {
        self.sum_timestamps += other.sum_timestamps;
        self.earliest_timestamp = min(self.earliest_timestamp, other.earliest_timestamp);
        self.num_values += other.num_values;
    }

    fn max_value_ms(&self, write_time: DateTime<Utc>) -> u64 {
        let write_time = write_time.timestamp_millis() as u64;
        write_time.saturating_sub(self.earliest_timestamp)
    }

    fn avg_value_ms(&self, write_time: DateTime<Utc>) -> u64 {
        let write_time = write_time.timestamp_millis() as u64;
        (write_time as f64 - (self.sum_timestamps / self.num_values as f64)) as u64
    }

    fn send_metric(&self, write_time: DateTime<Utc>, metric_name: &str) {
        if self.num_values == 0 {
            return;
        }

        timer!(
            format_args!("insertions.max_{}_ms", metric_name),
            self.max_value_ms(write_time)
        );

        timer!(
            format_args!("insertions.{}_ms", metric_name),
            self.avg_value_ms(write_time),
        );
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReplacementData {
    pub key: Vec<u8>, // Project id
    // Replacement message bytes. Newline-delimited series of json payloads
    pub value: Vec<u8>,
}

impl From<ReplacementData> for KafkaPayload {
    fn from(value: ReplacementData) -> KafkaPayload {
        KafkaPayload::new(Some(value.key), None, Some(value.value))
    }
}

impl From<KafkaPayload> for ReplacementData {
    fn from(value: KafkaPayload) -> ReplacementData {
        ReplacementData {
            key: value.key().unwrap().clone(),
            value: value.payload().unwrap().clone(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum InsertOrReplacement<T> {
    Insert(T),
    Replacement(ReplacementData),
}

/// The return value of message processors.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct InsertBatch {
    pub rows: RowData,
    pub origin_timestamp: Option<DateTime<Utc>>,
    pub sentry_received_timestamp: Option<DateTime<Utc>>,
    pub cogs_data: Option<CogsData>,
}

impl InsertBatch {
    pub fn from_rows<T>(
        rows: impl IntoIterator<Item = T>,
        origin_timestamp: Option<DateTime<Utc>>,
    ) -> anyhow::Result<Self>
    where
        T: Serialize,
    {
        let rows = RowData::from_rows(rows)?;
        Ok(Self {
            rows,
            origin_timestamp,
            sentry_received_timestamp: None,
            cogs_data: None,
        })
    }

    /// In case the processing function wants to skip the message, we return an empty batch.
    /// But instead of having the caller send an empty batch, lets make an explicit api for
    /// skipping. This way we can change the implementation later if we want to. Skipping ensures
    /// that the message is committed but not processed.
    pub fn skip() -> Self {
        Self::default()
    }
}

#[derive(Clone, Debug, Default)]
pub struct BytesInsertBatch<R> {
    rows: R,

    /// when the message was inserted into the snuba topic
    ///
    /// In Python this aggregate value is not explicitly tracked on BytesInsertBatch, as the batch
    /// type contains a list of the original kafka messages which all contain the individual
    /// timestamp values in metadata.
    message_timestamp: LatencyRecorder,

    /// when the event was received by Relay
    origin_timestamp: LatencyRecorder,

    /// when it was received by the ingest consumer in Sentry
    ///
    /// May not be recorded for some datasets. This is specifically used for metrics datasets, where
    /// this represents the latency from ingest-metrics (i.e. before the metrics indexer) to
    /// insertion into clickhouse
    sentry_received_timestamp: LatencyRecorder,

    // For each partition we store the offset and timestamp to be produced to the commit log
    commit_log_offsets: CommitLogOffsets,

    cogs_data: CogsData,
}

impl<R> BytesInsertBatch<R> {
    pub fn new(
        rows: R,
        message_timestamp: Option<DateTime<Utc>>,
        origin_timestamp: Option<DateTime<Utc>>,
        sentry_received_timestamp: Option<DateTime<Utc>>,
        commit_log_offsets: CommitLogOffsets,
        cogs_data: CogsData,
    ) -> Self {
        BytesInsertBatch {
            rows,
            message_timestamp: message_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            origin_timestamp: origin_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            sentry_received_timestamp: sentry_received_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            commit_log_offsets,
            cogs_data,
        }
    }

    pub fn commit_log_offsets(&self) -> &CommitLogOffsets {
        &self.commit_log_offsets
    }

    pub fn cogs_data(&self) -> &CogsData {
        &self.cogs_data
    }

    pub fn take(self) -> (R, BytesInsertBatch<()>) {
        let new = BytesInsertBatch {
            rows: (),
            message_timestamp: self.message_timestamp,
            origin_timestamp: self.origin_timestamp,
            sentry_received_timestamp: self.sentry_received_timestamp,
            commit_log_offsets: self.commit_log_offsets,
            cogs_data: self.cogs_data,
        };

        (self.rows, new)
    }

    pub fn record_message_latency(&self) {
        let write_time = Utc::now();

        self.message_timestamp.send_metric(write_time, "latency");
        self.origin_timestamp
            .send_metric(write_time, "end_to_end_latency");
        self.sentry_received_timestamp
            .send_metric(write_time, "sentry_received_latency");
    }
}

impl BytesInsertBatch<RowData> {
    pub fn len(&self) -> usize {
        self.rows.num_rows
    }
}

impl BytesInsertBatch<HttpBatch> {
    pub fn merge(mut self, other: BytesInsertBatch<RowData>) -> Self {
        self.rows
            .write_rows(&other.rows)
            .expect("failed to write rows to channel");
        self.commit_log_offsets.merge(other.commit_log_offsets);
        self.message_timestamp.merge(other.message_timestamp);
        self.origin_timestamp.merge(other.origin_timestamp);
        self.sentry_received_timestamp
            .merge(other.sentry_received_timestamp);
        self.cogs_data.merge(other.cogs_data);
        self
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, Default, PartialEq)]
pub struct RowData {
    pub encoded_rows: Vec<u8>,
    pub num_rows: usize,
}

impl RowData {
    pub fn from_rows<T>(rows: impl IntoIterator<Item = T>) -> anyhow::Result<Self>
    where
        T: Serialize,
    {
        let mut encoded_rows = Vec::new();
        let mut num_rows = 0;
        for row in rows {
            serde_json::to_writer(&mut encoded_rows, &row)?;
            debug_assert!(encoded_rows.ends_with(b"}"));
            encoded_rows.push(b'\n');
            num_rows += 1;
        }

        Ok(RowData {
            num_rows,
            encoded_rows,
        })
    }

    pub fn from_encoded_rows(rows: Vec<Vec<u8>>) -> Self {
        let mut encoded_rows = Vec::new();
        let mut num_rows = 0;
        for row in rows {
            encoded_rows.extend_from_slice(&row);
            encoded_rows.push(b'\n');
            num_rows += 1;
        }

        RowData {
            encoded_rows,
            num_rows,
        }
    }

    pub fn into_encoded_rows(self) -> Vec<u8> {
        self.encoded_rows
    }
}

#[derive(Clone, Debug)]
pub struct KafkaMessageMetadata {
    pub partition: u16,
    pub offset: u64,
    pub timestamp: DateTime<Utc>,
}

#[cfg(test)]
mod tests {
    use chrono::{TimeZone, Utc};

    use super::*;

    #[test]
    fn test_latency_recorder_basic() {
        // emulate how Reduce works when creating a batch for inserting into clickhouse, because
        // that's where we use it in practice. Reduce starts with a default() accumulator
        let mut accumulator = LatencyRecorder::default();

        // then there's one row being merged into the current batch
        let now = Utc.timestamp_opt(61, 0).unwrap();
        accumulator.merge(LatencyRecorder::from(now));

        // then another row
        let now = Utc.timestamp_opt(63, 0).unwrap();
        accumulator.merge(LatencyRecorder::from(now));

        // finally, when flushing the batch, we record the metric. let's see if LatencyRecorder did
        // the math right:
        assert_eq!(accumulator.num_values, 2);
        let now = Utc.timestamp_opt(65, 0).unwrap();
        assert_eq!(accumulator.max_value_ms(now), 4000);
        assert_eq!(accumulator.avg_value_ms(now), 3000);
    }
}
