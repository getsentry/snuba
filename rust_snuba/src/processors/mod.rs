mod functions;
mod profiles;
mod querylog;
mod spans;
mod utils;

use crate::types::{KafkaMessageMetadata, RowData};
use rust_arroyo::backends::kafka::types::KafkaPayload;

type ProcessingFunction = fn(KafkaPayload, KafkaMessageMetadata) -> anyhow::Result<RowData>;

pub fn get_processing_function(name: &str) -> Option<ProcessingFunction> {
    match name {
        "FunctionsMessageProcessor" => Some(functions::process_message),
        "ProfilesMessageProcessor" => Some(profiles::process_message),
        "QuerylogProcessor" => Some(querylog::process_message),
        "SpansMessageProcessor" => Some(spans::process_message),
        _ => None,
    }
}
