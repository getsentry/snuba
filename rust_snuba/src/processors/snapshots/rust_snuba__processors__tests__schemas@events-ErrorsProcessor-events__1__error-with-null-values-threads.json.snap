---
source: src/processors/mod.rs
description: "[\n  2,\n  \"insert\",\n  {\n    \"data\": {\n      \"event_id\": \"ced998ed44fa4cfc9211f48f248f3ce5\",\n      \"project\": 5428559,\n      \"release\": \"io.sentry.samples.android@6.16.0+2\",\n      \"dist\": \"2\",\n      \"platform\": \"java\",\n      \"message\": \"\",\n      \"datetime\": \"2023-03-17T13:34:59+00:00\",\n      \"tags\": [\n        [\"device\", \"sdk_gphone64_arm64\"],\n        [\"device.family\", \"sdk_gphone64_arm64\"],\n        [\"environment\", \"debug\"],\n        [\"isSideLoaded\", \"true\"],\n        [\"level\", \"error\"],\n        [\"os\", \"Android 13\"],\n        [\"os.name\", \"Android\"],\n        [\"os.rooted\", \"no\"],\n        [\"dist\", \"2\"],\n        [\"release\", \"io.sentry.samples.android@6.16.0+2\"],\n        [\"user\", \"id:b7478c1e-a9ba-4e5a-bc65-364142284f89\"],\n        [\"transaction\", \"MainActivity.capture_exception\"]\n      ],\n      \"_metrics\": {\n        \"bytes.ingested.event\": 22219,\n        \"bytes.ingested.event.attachment\": 125499,\n        \"bytes.stored.event\": 61436,\n        \"bytes.stored.event.attachment\": 125499\n      },\n      \"breadcrumbs\": {\n        \"values\": [\n          {\n            \"timestamp\": 1679060096.044,\n            \"type\": \"session\",\n            \"category\": \"app.lifecycle\",\n            \"level\": \"info\",\n            \"data\": {\n              \"state\": \"session.start\"\n            }\n          },\n          {\n            \"timestamp\": 1679060096.082,\n            \"type\": \"navigation\",\n            \"category\": \"ui.lifecycle\",\n            \"level\": \"info\",\n            \"data\": {\n              \"screen\": \"MainActivity\",\n              \"state\": \"created\"\n            }\n          },\n          {\n            \"timestamp\": 1679060096.138,\n            \"type\": \"navigation\",\n            \"category\": \"ui.lifecycle\",\n            \"level\": \"info\",\n            \"data\": {\n              \"screen\": \"MainActivity\",\n              \"state\": \"started\"\n            }\n          },\n          {\n            \"timestamp\": 1679060096.139,\n            \"type\": \"navigation\",\n            \"category\": \"app.lifecycle\",\n            \"level\": \"info\",\n            \"data\": {\n              \"state\": \"foreground\"\n            }\n          },\n          {\n            \"timestamp\": 1679060096.139,\n            \"type\": \"navigation\",\n            \"category\": \"ui.lifecycle\",\n            \"level\": \"info\",\n            \"data\": {\n              \"screen\": \"MainActivity\",\n              \"state\": \"resumed\"\n            }\n          },\n          {\n            \"timestamp\": 1679060099.782,\n            \"type\": \"user\",\n            \"category\": \"ui.click\",\n            \"level\": \"info\",\n            \"data\": {\n              \"view.class\": \"androidx.appcompat.widget.AppCompatButton\",\n              \"view.id\": \"capture_exception\"\n            }\n          }\n        ]\n      },\n      \"contexts\": {\n        \"app\": {\n          \"app_start_time\": \"2023-03-17T13:34:55.990Z\",\n          \"app_identifier\": \"io.sentry.samples.android\",\n          \"app_name\": \"Sentry sample\",\n          \"app_version\": \"6.16.0\",\n          \"app_build\": \"2\",\n          \"in_foreground\": true,\n          \"permissions\": {\n            \"ACCESS_NETWORK_STATE\": \"granted\",\n            \"CAMERA\": \"not_granted\",\n            \"FOREGROUND_SERVICE\": \"granted\",\n            \"INTERNET\": \"granted\",\n            \"READ_EXTERNAL_STORAGE\": \"not_granted\",\n            \"READ_PHONE_STATE\": \"not_granted\",\n            \"WRITE_EXTERNAL_STORAGE\": \"not_granted\"\n          },\n          \"type\": \"app\"\n        },\n        \"device\": {\n          \"name\": \"sdk_gphone64_arm64\",\n          \"family\": \"sdk_gphone64_arm64\",\n          \"model\": \"sdk_gphone64_arm64\",\n          \"model_id\": \"TE1A.220922.021\",\n          \"battery_level\": 100.0,\n          \"orientation\": \"portrait\",\n          \"manufacturer\": \"Google\",\n          \"brand\": \"google\",\n          \"screen_density\": 2.625,\n          \"screen_dpi\": 420,\n          \"online\": true,\n          \"charging\": false,\n          \"low_memory\": false,\n          \"simulator\": true,\n          \"memory_size\": 2057527296,\n          \"free_memory\": 641314816,\n          \"storage_size\": 6228115456,\n          \"free_storage\": 5105745920,\n          \"boot_time\": \"2023-03-17T11:15:57.308Z\",\n          \"timezone\": \"Europe/Vienna\",\n          \"archs\": [\"arm64-v8a\"],\n          \"battery_temperature\": 25.0,\n          \"connection_type\": \"wifi\",\n          \"id\": \"b7478c1e-a9ba-4e5a-bc65-364142284f89\",\n          \"language\": \"en\",\n          \"locale\": \"en_US\",\n          \"screen_height_pixels\": 2201,\n          \"screen_width_pixels\": 1080,\n          \"type\": \"device\"\n        },\n        \"os\": {\n          \"name\": \"Android\",\n          \"version\": \"13\",\n          \"build\": \"sdk_gphone64_arm64-userdebug 13 TE1A.220922.021 9526604 dev-keys\",\n          \"kernel_version\": \"5.15.41-android13-8-00055-g4f5025129fe8-ab8949913\",\n          \"rooted\": false,\n          \"type\": \"os\"\n        },\n        \"trace\": {\n          \"trace_id\": \"1da660e88c01430b85fe68537ed5838b\",\n          \"span_id\": \"d95788e2076d497a\",\n          \"op\": \"ui.action.click\",\n          \"status\": \"unknown\",\n          \"client_sample_rate\": 1.0,\n          \"type\": \"trace\"\n        }\n      },\n      \"culprit\": \"MainActivity.capture_exception\",\n      \"environment\": \"debug\",\n      \"exception\": {\n        \"values\": [\n          {\n            \"type\": \"Exception\",\n            \"value\": \"Some exception.\",\n            \"module\": \"java.lang\",\n            \"stacktrace\": {\n              \"frames\": [\n                {\n                  \"function\": \"main\",\n                  \"module\": \"com.android.internal.os.ZygoteInit\",\n                  \"filename\": \"ZygoteInit.java\",\n                  \"abs_path\": \"ZygoteInit.java\",\n                  \"lineno\": 936,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"run\",\n                  \"module\": \"com.android.internal.os.RuntimeInit$MethodAndArgsCaller\",\n                  \"filename\": \"RuntimeInit.java\",\n                  \"abs_path\": \"RuntimeInit.java\",\n                  \"lineno\": 548,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"invoke\",\n                  \"module\": \"java.lang.reflect.Method\",\n                  \"filename\": \"Method.java\",\n                  \"abs_path\": \"Method.java\",\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"main\",\n                  \"module\": \"android.app.ActivityThread\",\n                  \"filename\": \"ActivityThread.java\",\n                  \"abs_path\": \"ActivityThread.java\",\n                  \"lineno\": 7872,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"loop\",\n                  \"module\": \"android.os.Looper\",\n                  \"filename\": \"Looper.java\",\n                  \"abs_path\": \"Looper.java\",\n                  \"lineno\": 288,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"loopOnce\",\n                  \"module\": \"android.os.Looper\",\n                  \"filename\": \"Looper.java\",\n                  \"abs_path\": \"Looper.java\",\n                  \"lineno\": 201,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"dispatchMessage\",\n                  \"module\": \"android.os.Handler\",\n                  \"filename\": \"Handler.java\",\n                  \"abs_path\": \"Handler.java\",\n                  \"lineno\": 99,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"handleCallback\",\n                  \"module\": \"android.os.Handler\",\n                  \"filename\": \"Handler.java\",\n                  \"abs_path\": \"Handler.java\",\n                  \"lineno\": 942,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"run\",\n                  \"module\": \"android.view.View$PerformClick\",\n                  \"filename\": \"View.java\",\n                  \"abs_path\": \"View.java\",\n                  \"lineno\": 29334,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"-$$Nest$mperformClickInternal\",\n                  \"module\": \"android.view.View\",\n                  \"lineno\": 0,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"performClickInternal\",\n                  \"module\": \"android.view.View\",\n                  \"filename\": \"View.java\",\n                  \"abs_path\": \"View.java\",\n                  \"lineno\": 7483,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"performClick\",\n                  \"module\": \"android.view.View\",\n                  \"filename\": \"View.java\",\n                  \"abs_path\": \"View.java\",\n                  \"lineno\": 7506,\n                  \"in_app\": false,\n                  \"data\": {\n                    \"orig_in_app\": -1\n                  }\n                },\n                {\n                  \"function\": \"onClick\",\n                  \"module\": \"io.sentry.samples.android.MainActivity$$ExternalSyntheticLambda7\",\n                  \"lineno\": 0,\n                  \"in_app\": true\n                },\n                {\n                  \"function\": \"lambda$onCreate$6\",\n                  \"module\": \"io.sentry.samples.android.MainActivity\",\n                  \"filename\": \"MainActivity.java\",\n                  \"abs_path\": \"MainActivity.java\",\n                  \"lineno\": 107,\n                  \"in_app\": true\n                }\n              ]\n            },\n            \"thread_id\": 2\n          }\n        ]\n      },\n      \"fingerprint\": [\"{{ default }}\"],\n      \"grouping_config\": {\n        \"enhancements\": \"eJybzDRxY3J-bm5-npWRgaGlroGxrpHxxEkT1-Zm5usVp-aVFFXqaWlNZAQAKGsOFg\",\n        \"id\": \"newstyle:2019-10-29\"\n      },\n      \"hashes\": [\n        \"7956a3c218a27566572b64cad9f2b3e7\",\n        \"507662f072d47719837844a6f3f579c4\"\n      ],\n      \"ingest_path\": [\n        {\n          \"version\": \"23.3.0\",\n          \"public_key\": \"XE7QiyuNlja9PZ7I9qJlwQotzecWrUIN91BAO7Q5R38\"\n        }\n      ],\n      \"key_id\": \"1336851\",\n      \"level\": \"error\",\n      \"location\": \"MainActivity.java\",\n      \"logger\": \"\",\n      \"metadata\": {\n        \"display_title_with_tree_label\": false,\n        \"filename\": \"MainActivity.java\",\n        \"function\": \"lambda$onCreate$6\",\n        \"type\": \"Exception\",\n        \"value\": \"Some exception.\"\n      },\n      \"nodestore_insert\": 1679060109.558613,\n      \"received\": 1679060103.68783,\n      \"sdk\": {\n        \"name\": \"sentry.java.android\",\n        \"version\": \"6.16.0\",\n        \"integrations\": [\n          \"ComposeUserInteraction\",\n          \"UncaughtExceptionHandler\",\n          \"ShutdownHook\",\n          \"Ndk\",\n          \"AppLifecycle\",\n          \"Anr\",\n          \"ActivityLifecycle\",\n          \"UserInteraction\",\n          \"FragmentLifecycle\",\n          \"Timber\",\n          \"AppComponentsBreadcrumbs\",\n          \"SystemEventsBreadcrumbs\",\n          \"TempSensorBreadcrumbs\"\n        ],\n        \"packages\": [\n          {\n            \"name\": \"maven:io.sentry:sentry\",\n            \"version\": \"6.16.0\"\n          },\n          {\n            \"name\": \"maven:io.sentry:sentry-android-core\",\n            \"version\": \"6.16.0\"\n          },\n          {\n            \"name\": \"maven:io.sentry:sentry-compose\",\n            \"version\": \"6.16.0\"\n          },\n          {\n            \"name\": \"maven:io.sentry:sentry-android-ndk\",\n            \"version\": \"6.16.0\"\n          },\n          {\n            \"name\": \"maven:io.sentry:sentry-android-fragment\",\n            \"version\": \"6.16.0\"\n          },\n          {\n            \"name\": \"maven:io.sentry:sentry-android-timber\",\n            \"version\": \"6.16.0\"\n          }\n        ]\n      },\n      \"threads\": {\n        \"values\": null\n      },\n      \"timestamp\": 1679060099.815,\n      \"title\": \"Exception: Some exception.\",\n      \"transaction\": \"MainActivity.capture_exception\",\n      \"type\": \"error\",\n      \"user\": {\n        \"id\": \"b7478c1e-a9ba-4e5a-bc65-364142284f89\",\n        \"ip_address\": \"84.115.213.8\",\n        \"geo\": {\n          \"country_code\": \"AT\",\n          \"city\": \"Vienna\",\n          \"region\": \"Austria\"\n        }\n      },\n      \"version\": \"7\"\n    },\n    \"datetime\": \"2023-02-27T15:40:12.223000Z\",\n    \"event_id\": \"9cdc4c32dff14fbbb012b0aa9e908126\",\n    \"group_id\": 124,\n    \"group_ids\": [124],\n    \"message\": \"hello world\",\n    \"organization_id\": 123,\n    \"platform\": \"javascript\",\n    \"primary_hash\": \"061cf02b26374d108694d6643a7a2f4e\",\n    \"project_id\": 6036610\n  },\n  {\n    \"group_states\": [\n      {\n        \"id\": \"124\",\n        \"is_new\": false,\n        \"is_new_group_environment\": false,\n        \"is_regression\": false\n      }\n    ],\n    \"is_new\": false,\n    \"is_new_group_environment\": false,\n    \"is_regression\": false,\n    \"queue\": \"post_process_errors\",\n    \"skip_consume\": false\n  }\n]\n"
expression: snapshot_payload
---
[
  {
    "contexts.key": [
      "app.app_build",
      "app.app_identifier",
      "app.app_name",
      "app.app_start_time",
      "app.app_version",
      "app.in_foreground",
      "device.battery_level",
      "device.battery_temperature",
      "device.boot_time",
      "device.brand",
      "device.charging",
      "device.connection_type",
      "device.family",
      "device.free_memory",
      "device.free_storage",
      "device.id",
      "device.language",
      "device.locale",
      "device.low_memory",
      "device.manufacturer",
      "device.memory_size",
      "device.model",
      "device.model_id",
      "device.name",
      "device.online",
      "device.orientation",
      "device.screen_density",
      "device.screen_dpi",
      "device.screen_height_pixels",
      "device.screen_width_pixels",
      "device.simulator",
      "device.storage_size",
      "device.timezone",
      "geo.city",
      "geo.country_code",
      "geo.region",
      "os.build",
      "os.kernel_version",
      "os.name",
      "os.rooted",
      "os.version",
      "trace.client_sample_rate",
      "trace.op",
      "trace.span_id",
      "trace.status",
      "trace.trace_id"
    ],
    "contexts.value": [
      "2",
      "io.sentry.samples.android",
      "Sentry sample",
      "2023-03-17T13:34:55.990Z",
      "6.16.0",
      "True",
      "100.0",
      "25.0",
      "2023-03-17T11:15:57.308Z",
      "google",
      "False",
      "wifi",
      "sdk_gphone64_arm64",
      "641314816",
      "5105745920",
      "b7478c1e-a9ba-4e5a-bc65-364142284f89",
      "en",
      "en_US",
      "False",
      "Google",
      "2057527296",
      "sdk_gphone64_arm64",
      "TE1A.220922.021",
      "sdk_gphone64_arm64",
      "True",
      "portrait",
      "2.625",
      "420",
      "2201",
      "1080",
      "True",
      "6228115456",
      "Europe/Vienna",
      "Vienna",
      "AT",
      "Austria",
      "sdk_gphone64_arm64-userdebug 13 TE1A.220922.021 9526604 dev-keys",
      "5.15.41-android13-8-00055-g4f5025129fe8-ab8949913",
      "Android",
      "False",
      "13",
      "1.0",
      "ui.action.click",
      "d95788e2076d497a",
      "unknown",
      "1da660e88c01430b85fe68537ed5838b"
    ],
    "culprit": "MainActivity.capture_exception",
    "deleted": 0,
    "dist": null,
    "environment": "debug",
    "event_id": "9cdc4c32-dff1-4fbb-b012-b0aa9e908126",
    "exception_frames.abs_path": [
      "ZygoteInit.java",
      "RuntimeInit.java",
      "Method.java",
      "ActivityThread.java",
      "Looper.java",
      "Looper.java",
      "Handler.java",
      "Handler.java",
      "View.java",
      null,
      "View.java",
      "View.java",
      null,
      "MainActivity.java"
    ],
    "exception_frames.colno": [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null
    ],
    "exception_frames.filename": [
      "ZygoteInit.java",
      "RuntimeInit.java",
      "Method.java",
      "ActivityThread.java",
      "Looper.java",
      "Looper.java",
      "Handler.java",
      "Handler.java",
      "View.java",
      null,
      "View.java",
      "View.java",
      null,
      "MainActivity.java"
    ],
    "exception_frames.function": [
      "main",
      "run",
      "invoke",
      "main",
      "loop",
      "loopOnce",
      "dispatchMessage",
      "handleCallback",
      "run",
      "-$$Nest$mperformClickInternal",
      "performClickInternal",
      "performClick",
      "onClick",
      "lambda$onCreate$6"
    ],
    "exception_frames.in_app": [
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      false,
      true,
      true
    ],
    "exception_frames.lineno": [
      936,
      548,
      null,
      7872,
      288,
      201,
      99,
      942,
      29334,
      0,
      7483,
      7506,
      0,
      107
    ],
    "exception_frames.module": [
      "com.android.internal.os.ZygoteInit",
      "com.android.internal.os.RuntimeInit$MethodAndArgsCaller",
      "java.lang.reflect.Method",
      "android.app.ActivityThread",
      "android.os.Looper",
      "android.os.Looper",
      "android.os.Handler",
      "android.os.Handler",
      "android.view.View$PerformClick",
      "android.view.View",
      "android.view.View",
      "android.view.View",
      "io.sentry.samples.android.MainActivity$$ExternalSyntheticLambda7",
      "io.sentry.samples.android.MainActivity"
    ],
    "exception_frames.package": [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null
    ],
    "exception_frames.stack_level": [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0
    ],
    "exception_stacks.mechanism_handled": [
      null
    ],
    "exception_stacks.mechanism_type": [
      null
    ],
    "exception_stacks.type": [
      "Exception"
    ],
    "exception_stacks.value": [
      "Some exception."
    ],
    "group_id": 124,
    "hierarchical_hashes": [],
    "http_method": null,
    "http_referer": null,
    "ip_address_v4": "84.115.213.8",
    "level": "error",
    "location": "MainActivity.java",
    "message": "hello world",
    "message_timestamp": "<event timestamp>",
    "modules.name": [],
    "modules.version": [],
    "num_processing_errors": 0,
    "offset": 1,
    "partition": 0,
    "platform": "javascript",
    "primary_hash": "061cf02b-2637-4d10-8694-d6643a7a2f4e",
    "project_id": 6036610,
    "received": 1679060103,
    "release": null,
    "retention_days": 90,
    "sdk_integrations": [
      "ComposeUserInteraction",
      "UncaughtExceptionHandler",
      "ShutdownHook",
      "Ndk",
      "AppLifecycle",
      "Anr",
      "ActivityLifecycle",
      "UserInteraction",
      "FragmentLifecycle",
      "Timber",
      "AppComponentsBreadcrumbs",
      "SystemEventsBreadcrumbs",
      "TempSensorBreadcrumbs"
    ],
    "sdk_name": "sentry.java.android",
    "sdk_version": "6.16.0",
    "span_id": 15661136733572778362,
    "tags.key": [
      "device",
      "device.family",
      "dist",
      "environment",
      "isSideLoaded",
      "level",
      "os",
      "os.name",
      "os.rooted",
      "release",
      "transaction",
      "user"
    ],
    "tags.value": [
      "sdk_gphone64_arm64",
      "sdk_gphone64_arm64",
      "2",
      "debug",
      "true",
      "error",
      "Android 13",
      "Android",
      "no",
      "io.sentry.samples.android@6.16.0+2",
      "MainActivity.capture_exception",
      "id:b7478c1e-a9ba-4e5a-bc65-364142284f89"
    ],
    "timestamp": 1677512412,
    "title": "Exception: Some exception.",
    "trace_id": "1da660e8-8c01-430b-85fe-68537ed5838b",
    "transaction_name": "MainActivity.capture_exception",
    "type": "error",
    "user": "",
    "user_email": null,
    "user_id": "b7478c1e-a9ba-4e5a-bc65-364142284f89",
    "user_name": null,
    "version": "7"
  }
]
